File: dart/lib/src/nepali_calendar.dart
/// Jhigu Patro Utils — Core Calendar Engine (Dart)
///
/// Converts between Gregorian (AD) and Bikram Sambat (BS/Nepali) dates.
/// Zero external dependencies. Works in Flutter, Dart CLI, and server Dart.
///
/// Setup (call once before using conversion functions):
/// ```dart
/// NepaliCalendar.initialize(InlineCalendarDataSource(years: ..., ...));
/// ```
library;

import 'types.dart';

// ============================================================================
// CONSTANTS
// ============================================================================

const int _bsBaseYear = 2000;

/// BS 2000/1/1 = AD 1943-04-14
final DateTime _bsBaseDate = DateTime.utc(1943, 4, 14);

const Duration _nepalOffset = Duration(hours: 5, minutes: 45);

// ============================================================================
// NAMES
// ============================================================================

const List<String> nepaliMonthsEn = [
  'Baisakh', 'Jestha', 'Asar', 'Shrawan', 'Bhadra', 'Ashwin',
  'Kartik', 'Mangsir', 'Poush', 'Magh', 'Falgun', 'Chaitra',
];

const List<String> nepaliMonthsNe = [
  'बैशाख', 'जेठ', 'असार', 'साउन', 'भदौ', 'असोज',
  'कार्तिक', 'मंसिर', 'पौष', 'माघ', 'फागुन', 'चैत्र',
];

const List<String> englishMonths = [
  'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December',
];

/// Sunday = index 0 (matches DateTime.sunday = 7, handled below).
const List<String> nepaliDaysEn = [
  'Aaitabar', 'Sombar', 'Mangalbar', 'Budhabar', 'Bihibar', 'Shukrabar', 'Shanibar',
];

const List<String> nepaliDaysNe = [
  'आइतबार', 'सोमबार', 'मंगलबार', 'बुधबार', 'बिहिबार', 'शुक्रबार', 'शनिबार',
];

const List<String> nepaliDigits = ['०', '१', '२', '३', '४', '५', '६', '७', '८', '९'];

// ============================================================================
// DATA SOURCE INJECTION
// ============================================================================

class NepaliCalendar {
  NepaliCalendar._();

  static NepaliCalendarDataSource? _source;

  /// Register the data source. Must be called once before any conversion.
  static void initialize(NepaliCalendarDataSource source) {
    _source = source;
  }

  static NepaliCalendarDataSource get _ds {
    if (_source == null) {
      throw StateError(
        'jhigu_patro_utils: calendar not initialized. Call NepaliCalendar.initialize() first.',
      );
    }
    return _source!;
  }

  // ============================================================================
  // DATA HELPERS
  // ============================================================================

  /// Days in [month] (1-indexed) of BS [year].
  static int getDaysInBSMonth(int year, int month) {
    final data = _ds.getYearData(year);
    if (data == null) throw ArgumentError('No calendar data for BS year $year');
    return data[month - 1];
  }

  /// Total days in BS [year].
  static int getTotalDaysInBSYear(int year) {
    final data = _ds.getYearData(year);
    if (data == null) throw ArgumentError('No calendar data for BS year $year');
    return data.fold(0, (s, d) => s + d);
  }

  /// Returns true if the BS date is within range and valid.
  static bool isValidBSDate(int year, int month, int day) {
    if (month < 1 || month > 12 || day < 1) return false;
    final data = _ds.getYearData(year);
    if (data == null) return false;
    return day <= data[month - 1];
  }

  /// All supported BS years.
  static List<int> getAvailableBSYears() {
    final src = _ds;
    return [
      for (int y = src.supportedStartYear; y <= src.supportedEndYear; y++)
        if (src.getYearData(y) != null) y,
    ];
  }

  // ============================================================================
  // CONVERSION
  // ============================================================================

  /// Convert a Gregorian [DateTime] (AD) to a [NepaliDate] (BS).
  static NepaliDate convertADToBS(DateTime adDate) {
    final adUtc = DateTime.utc(adDate.year, adDate.month, adDate.day);
    int remaining = adUtc.difference(_bsBaseDate).inDays;

    int year = _bsBaseYear;
    int month = 1;

    while (remaining > 0) {
      final diy = getTotalDaysInBSYear(year);
      if (remaining >= diy) {
        remaining -= diy;
        year++;
      } else {
        break;
      }
    }

    while (remaining > 0) {
      final dim = getDaysInBSMonth(year, month);
      if (remaining >= dim) {
        remaining -= dim;
        month++;
      } else {
        break;
      }
    }

    return NepaliDate(year: year, month: month, day: remaining + 1);
  }

  /// Convert a BS date to a Gregorian [DateTime] (AD).
  static DateTime convertBSToAD(int bsYear, int bsMonth, int bsDay) {
    if (!isValidBSDate(bsYear, bsMonth, bsDay)) {
      throw ArgumentError('Invalid BS date: $bsYear/$bsMonth/$bsDay');
    }

    int total = 0;
    for (int y = _bsBaseYear; y < bsYear; y++) total += getTotalDaysInBSYear(y);
    for (int m = 1; m < bsMonth; m++) total += getDaysInBSMonth(bsYear, m);
    total += bsDay - 1;

    return _bsBaseDate.add(Duration(days: total));
  }

  // ============================================================================
  // TIMEZONE
  // ============================================================================

  /// Convert a UTC [DateTime] to Nepal time (GMT+5:45).
  static DateTime toNepalTime(DateTime utc) => utc.toUtc().add(_nepalOffset);

  /// Reverse of [toNepalTime].
  static DateTime fromNepalTime(DateTime nepDate) => nepDate.subtract(_nepalOffset);

  /// Current Nepal local date as a [NepaliDate].
  static NepaliDate getCurrentNepaliDate() {
    final nep = toNepalTime(DateTime.now().toUtc());
    return convertADToBS(DateTime.utc(nep.year, nep.month, nep.day));
  }

  // ============================================================================
  // FORMATTING
  // ============================================================================

  /// Convert an integer to Nepali (Devanagari) digit string.
  static String toNepaliDigits(int num) {
    return num.toString().split('').map((ch) {
      final d = int.tryParse(ch);
      return d != null ? nepaliDigits[d] : ch;
    }).join();
  }

  /// Day-of-week index (0=Sunday) from a [DateTime].
  static int _dayIndex(DateTime date) => date.weekday % 7;

  /// Weekday name for [date].
  static String getNepaliDayName(DateTime date, {bool nepali = false}) {
    final days = nepali ? nepaliDaysNe : nepaliDaysEn;
    return days[_dayIndex(date)];
  }

  /// Nepali month name (1-indexed).
  static String getNepaliMonthName(int month, {bool nepali = false}) {
    final months = nepali ? nepaliMonthsNe : nepaliMonthsEn;
    return months[month - 1];
  }

  /// Format a [NepaliDate] as a readable string.
  /// English: "Falgun 8, 2082"  Nepali: "फागुन 8, 2082"
  static String formatNepaliDate(NepaliDate date, {bool nepali = false}) {
    final month = getNepaliMonthName(date.month, nepali: nepali);
    return '$month ${date.day}, ${date.year}';
  }

  /// Format a Gregorian [DateTime] as "April 14, 2024".
  static String formatEnglishDate(DateTime date) =>
      '${englishMonths[date.month - 1]} ${date.day}, ${date.year}';

  /// Format a BS date with weekday name.
  /// Example: "८ फागुन २०८२, शुक्रबार"
  static String formatNepaliDateWithDay(NepaliDate nepaliDate, {DateTime? adDate}) {
    final ad = adDate ?? convertBSToAD(nepaliDate.year, nepaliDate.month, nepaliDate.day);
    final day = toNepaliDigits(nepaliDate.day);
    final month = nepaliMonthsNe[nepaliDate.month - 1];
    final year = toNepaliDigits(nepaliDate.year);
    final weekday = getNepaliDayName(ad, nepali: true);
    return '$day $month $year, $weekday';
  }

  /// Current Nepali date formatted with day name in Nepal time.
  static String getCurrentNepaliDateFormatted() {
    final nep = toNepalTime(DateTime.now().toUtc());
    final local = DateTime.utc(nep.year, nep.month, nep.day);
    return formatNepaliDateWithDay(convertADToBS(local), adDate: local);
  }

  /// Format a [DateTime] in Nepal timezone as "YYYY-MM-DD HH:mm".
  static String formatNepalDateTime(DateTime? date) {
    if (date == null) return '';
    final nep = toNepalTime(date.toUtc());
    final y = nep.year.toString().padLeft(4, '0');
    final mo = nep.month.toString().padLeft(2, '0');
    final d = nep.day.toString().padLeft(2, '0');
    final h = nep.hour.toString().padLeft(2, '0');
    final mi = nep.minute.toString().padLeft(2, '0');
    return '$y-$mo-$d $h:$mi';
  }

  /// Relative time string in Nepali (e.g. "१५ मिनेट अगाडि").
  /// Past midnight Nepal time → shows the BS date string.
  static String getRelativeTimeInNepali(DateTime? date) {
    if (date == null) return 'हालै';

    final nowNep = toNepalTime(DateTime.now().toUtc());
    final dateNep = toNepalTime(date.toUtc());

    final todayMidnight = DateTime.utc(nowNep.year, nowNep.month, nowNep.day);

    if (dateNep.isBefore(todayMidnight)) {
      return _bsDateString(convertADToBS(dateNep));
    }

    final diff = nowNep.difference(dateNep);
    if (diff.inSeconds < 1) return 'हालै';

    if (diff.inSeconds < 60) return '${toNepaliDigits(diff.inSeconds)} सेकेण्ड अगाडि';
    if (diff.inMinutes < 60) return '${toNepaliDigits(diff.inMinutes)} मिनेट अगाडि';
    if (diff.inHours < 24)   return '${toNepaliDigits(diff.inHours)} घण्टा अगाडि';
    return _bsDateString(convertADToBS(dateNep));
  }

  static String _bsDateString(NepaliDate nd) =>
      '${toNepaliDigits(nd.day)} ${nepaliMonthsNe[nd.month - 1]} ${toNepaliDigits(nd.year)}';
}

File: ts/src/nepaliCalendar.ts
/**
 * Jhigu Patro Utils — Core Calendar Engine
 *
 * Converts between Gregorian (AD) and Bikram Sambat (BS/Nepali) dates.
 * Zero external dependencies. Works in browser, Node.js, Deno, Bun, and any
 * JS/TS runtime.
 *
 * Key facts:
 *   - BS year 2000 Baisakh 1 = AD April 14, 1943
 *   - BS is ~56–57 years ahead of AD
 *   - Each BS month has 29–32 days (varies per year, stored in data source)
 *   - Nepal timezone: GMT+5:45
 *
 * Setup (call once before using conversion functions):
 *   const source = new InlineCalendarDataSource(years, startYear, endYear);
 *   initializeCalendar(source);
 */

import type { NepaliCalendarDataSource, NepaliDate } from './types.js';

// ============================================================================
// CONSTANTS
// ============================================================================

const BS_BASE_YEAR = 2000;
// BS 2000/1/1 = AD 1943 April 14 (month 0-indexed in JS Date)
const BS_BASE_DATE = new Date(1943, 3, 14);
const BASE_UTC = Date.UTC(1943, 3, 14);
const NEPAL_OFFSET_MS = (5 * 60 + 45) * 60 * 1000; // GMT+5:45

// ============================================================================
// MONTH / DAY / DIGIT NAMES
// ============================================================================

export const NEPALI_MONTHS_EN = [
  'Baisakh', 'Jestha', 'Asar', 'Shrawan', 'Bhadra', 'Ashwin',
  'Kartik', 'Mangsir', 'Poush', 'Magh', 'Falgun', 'Chaitra',
] as const;

export const NEPALI_MONTHS_NE = [
  'बैशाख', 'जेठ', 'असार', 'साउन', 'भदौ', 'असोज',
  'कार्तिक', 'मंसिर', 'पौष', 'माघ', 'फागुन', 'चैत्र',
] as const;

export const ENGLISH_MONTHS = [
  'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December',
] as const;

export const NEPALI_DAYS_EN = [
  'Aaitabar', 'Sombar', 'Mangalbar', 'Budhabar', 'Bihibar', 'Shukrabar', 'Shanibar',
] as const;

export const NEPALI_DAYS_NE = [
  'आइतबार', 'सोमबार', 'मंगलबार', 'बुधबार', 'बिहिबार', 'शुक्रबार', 'शनिबार',
] as const;

export const NEPALI_DIGITS = ['०', '१', '२', '३', '४', '५', '६', '७', '८', '९'] as const;

// ============================================================================
// DATA SOURCE INJECTION
// ============================================================================

let _dataSource: NepaliCalendarDataSource | null = null;

/**
 * Register the data source. Must be called once before any conversion.
 */
export function initializeCalendar(source: NepaliCalendarDataSource): void {
  _dataSource = source;
}

function ds(): NepaliCalendarDataSource {
  if (!_dataSource) {
    throw new Error(
      'jhigu-patro-utils: calendar not initialized. Call initializeCalendar() first.'
    );
  }
  return _dataSource;
}

// ============================================================================
// DATA HELPERS
// ============================================================================

/** Days in a given BS month (month is 1-indexed). */
export function getDaysInBSMonth(year: number, month: number): number {
  const yearData = ds().getYearData(year);
  if (!yearData) throw new Error(`No calendar data for BS year ${year}`);
  return yearData[month - 1];
}

/** Total days in a given BS year. */
export function getTotalDaysInBSYear(year: number): number {
  const yearData = ds().getYearData(year);
  if (!yearData) throw new Error(`No calendar data for BS year ${year}`);
  return yearData.reduce((sum, d) => sum + d, 0);
}

/** Returns true if the given BS date is within supported range and valid. */
export function isValidBSDate(year: number, month: number, day: number): boolean {
  if (month < 1 || month > 12 || day < 1) return false;
  const yearData = ds().getYearData(year);
  if (!yearData) return false;
  return day <= yearData[month - 1];
}

/** All supported BS years from the loaded data source. */
export function getAvailableBSYears(): number[] {
  const src = ds();
  const years: number[] = [];
  for (let y = src.supportedStartYear; y <= src.supportedEndYear; y++) {
    if (src.getYearData(y)) years.push(y);
  }
  return years;
}

/** AD year range spanned by the loaded BS data. */
export function getAvailableADYears(): number[] {
  const bsYears = getAvailableBSYears();
  if (bsYears.length === 0) return [];
  const minAD = convertBSToAD(bsYears[0], 1, 1);
  const maxBS = bsYears[bsYears.length - 1];
  const maxAD = convertBSToAD(maxBS, 12, getDaysInBSMonth(maxBS, 12));
  const years: number[] = [];
  for (let y = minAD.getFullYear(); y <= maxAD.getFullYear(); y++) years.push(y);
  return years;
}

// ============================================================================
// CONVERSION
// ============================================================================

/** Convert a JS Date (AD) to a Nepali (BS) date. */
export function convertADToBS(adDate: Date): NepaliDate {
  const adUTC = Date.UTC(adDate.getFullYear(), adDate.getMonth(), adDate.getDate());
  let remaining = Math.round((adUTC - BASE_UTC) / 86_400_000);

  let year = BS_BASE_YEAR;
  let month = 1;

  while (remaining > 0) {
    const diy = getTotalDaysInBSYear(year);
    if (remaining >= diy) { remaining -= diy; year++; }
    else break;
  }

  while (remaining > 0) {
    const dim = getDaysInBSMonth(year, month);
    if (remaining >= dim) { remaining -= dim; month++; }
    else break;
  }

  return { year, month, day: remaining + 1 };
}

/** Convert a BS date to a JS Date (AD). */
export function convertBSToAD(bsYear: number, bsMonth: number, bsDay: number): Date {
  if (!isValidBSDate(bsYear, bsMonth, bsDay)) {
    throw new Error(`Invalid BS date: ${bsYear}/${bsMonth}/${bsDay}`);
  }

  let total = 0;
  for (let y = BS_BASE_YEAR; y < bsYear; y++) total += getTotalDaysInBSYear(y);
  for (let m = 1; m < bsMonth; m++) total += getDaysInBSMonth(bsYear, m);
  total += bsDay - 1;

  const result = new Date(BS_BASE_DATE);
  result.setDate(result.getDate() + total);
  return result;
}

// ============================================================================
// TIMEZONE
// ============================================================================

/** Shift any date/timestamp to Nepal time (GMT+5:45) encoded as a UTC Date. */
export function toNepalTime(date: Date | string): Date {
  const d = typeof date === 'string' ? new Date(date) : date;
  return new Date(d.getTime() + NEPAL_OFFSET_MS);
}

/** Reverse of toNepalTime — convert Nepal-encoded date back to UTC. */
export function fromNepalTime(date: Date): Date {
  return new Date(date.getTime() - NEPAL_OFFSET_MS);
}

/** Current Nepal local date as a BS date. Timezone-safe regardless of server locale. */
export function getCurrentNepaliDate(): NepaliDate {
  const nep = toNepalTime(new Date());
  return convertADToBS(new Date(nep.getUTCFullYear(), nep.getUTCMonth(), nep.getUTCDate()));
}

// ============================================================================
// FORMATTING
// ============================================================================

/** Convert an integer to Nepali (Devanagari) digit string. */
export function toNepaliDigits(num: number): string {
  return String(num).split('').map(ch => {
    const d = parseInt(ch, 10);
    return isNaN(d) ? ch : NEPALI_DIGITS[d];
  }).join('');
}

/** Format a BS date as a readable string. */
export function formatNepaliDate(date: NepaliDate, locale: 'en' | 'ne' = 'en'): string {
  const months = locale === 'ne' ? NEPALI_MONTHS_NE : NEPALI_MONTHS_EN;
  return `${months[date.month - 1]} ${date.day}, ${date.year}`;
}

/** Format an AD Date as a readable string. */
export function formatEnglishDate(date: Date): string {
  return `${ENGLISH_MONTHS[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
}

/** Get the Nepali month name (1-indexed). */
export function getNepaliMonthName(month: number, locale: 'en' | 'ne' = 'en'): string {
  const months = locale === 'ne' ? NEPALI_MONTHS_NE : NEPALI_MONTHS_EN;
  return months[month - 1] ?? '';
}

/** Get day-of-week name for a JS Date. */
export function getNepaliDayName(date: Date, locale: 'en' | 'ne' = 'en'): string {
  const days = locale === 'ne' ? NEPALI_DAYS_NE : NEPALI_DAYS_EN;
  return days[date.getDay()];
}

/**
 * Format a BS date with Nepali day name.
 * Example output: "८ फागुन २०८२, शुक्रबार"
 */
export function formatNepaliDateWithDay(nepaliDate: NepaliDate, adDate?: Date): string {
  const ad = adDate ?? convertBSToAD(nepaliDate.year, nepaliDate.month, nepaliDate.day);
  const day = toNepaliDigits(nepaliDate.day);
  const month = NEPALI_MONTHS_NE[nepaliDate.month - 1];
  const year = toNepaliDigits(nepaliDate.year);
  const weekday = getNepaliDayName(ad, 'ne');
  return `${day} ${month} ${year}, ${weekday}`;
}

/**
 * Current Nepali date formatted with day name in Nepal time.
 * Example: "८ फागुन २०८२, शुक्रबार"
 */
export function getCurrentNepaliDateFormatted(): string {
  const nep = toNepalTime(new Date());
  const localDate = new Date(nep.getUTCFullYear(), nep.getUTCMonth(), nep.getUTCDate());
  return formatNepaliDateWithDay(convertADToBS(localDate), localDate);
}

/**
 * Format date in Nepal timezone as "YYYY-MM-DD HH:mm".
 */
export function formatNepalDateTime(date: Date | string | null | undefined): string {
  if (!date) return '';
  try {
    const d = toNepalTime(typeof date === 'string' ? new Date(date) : date);
    const y = d.getUTCFullYear();
    const mo = String(d.getUTCMonth() + 1).padStart(2, '0');
    const dy = String(d.getUTCDate()).padStart(2, '0');
    const h = String(d.getUTCHours()).padStart(2, '0');
    const mi = String(d.getUTCMinutes()).padStart(2, '0');
    return `${y}-${mo}-${dy} ${h}:${mi}`;
  } catch {
    return '';
  }
}

/**
 * Relative time string in Nepali (e.g. "१५ मिनेट अगाडि").
 * Once the date is past midnight Nepal time, shows the BS date instead.
 */
export function getRelativeTimeInNepali(date: Date | string | null | undefined): string {
  if (!date) return 'हालै';
  try {
    const d = typeof date === 'string' ? new Date(date) : date;
    if (isNaN(d.getTime())) return 'हालै';

    const nowNep = toNepalTime(new Date());
    const dateNep = toNepalTime(d);

    const todayMidnight = new Date(nowNep);
    todayMidnight.setUTCHours(0, 0, 0, 0);

    if (dateNep < todayMidnight) {
      return _bsDateString(convertADToBS(dateNep));
    }

    const diffMs = nowNep.getTime() - dateNep.getTime();
    if (diffMs < 1000) return 'हालै';

    const sec = Math.floor(diffMs / 1000);
    const min = Math.floor(sec / 60);
    const hr = Math.floor(min / 60);

    if (sec < 60) return `${toNepaliDigits(sec)} सेकेण्ड अगाडि`;
    if (min < 60) return `${toNepaliDigits(min)} मिनेट अगाडि`;
    if (hr < 24)  return `${toNepaliDigits(hr)} घण्टा अगाडि`;
    return _bsDateString(convertADToBS(dateNep));
  } catch {
    return 'हालै';
  }
}

function _bsDateString(nd: NepaliDate): string {
  return `${toNepaliDigits(nd.day)} ${NEPALI_MONTHS_NE[nd.month - 1]} ${toNepaliDigits(nd.year)}`;
}
